app/services/sync_canvas_grades.rb

```
@@ -0,0 +1,111 @@
+# frozen_string_literal: true
+
+require 'canvas_api'
+require 'salesforce_api'
+
+class SyncCanvasGrades
+
+  def initialize
+  end
+
+  def run
+    program_data = SalesforceAPI.client.get_current_and_future_accelerator_programs
+    canvas_course_ids = program_data['records'].map { |p| p['Canvas_Cloud_Accelerator_Course_ID__c'] }.uniq.compact
+    Honeycomb.add_field('sync_canvas_grades.courses.count', canvas_course_ids.count)
+    Rails.logger.info("Found #{canvas_course_ids.count} courses")
+
+    canvas_course_ids.each do |canvas_course_id|
+      Honeycomb.start_span(name: 'sync_submissions') do
+        Honeycomb.add_field('canvas.course.id', canvas_course_id.to_s)
+        # Sync grades.
+        Rails.logger.info("Syncing grades for canvas_course_id=#{canvas_course_id}")
+        submissions = CanvasAPI.client.get_submission_data(canvas_course_id)
```

We should keep an eye on how resource intensive this is. I'm assuming if there are file uploads, it doesn't actually have the file contents but rather a link to the file or something, but I'm guessing that the entirety of the submission data for a course could get pretty big. Actually, should we send a Honeycomb field with the size of this string/hash?

---

Also, if one course fails to sync due to the API throwing, do we want to keep trying the others or fail the whole sync? I assume fail the whole sync like you've written it until we see that actually happening in the wild and it's for a reason where we actually do want to just log that error but try the rest of the courses.

---

Added the count of each of the record types (submissions, rubric_assessments, rubrics) to HC. Though we'll never receive that count if it crashes before it can count them. Since the API response is paginated, we can also count the number of requests to see how many pages there were. There is a risk the number of submissions could get overwhelming in a real course, and overflow whatever the max variable size is in Ruby, but I thought that was something to worry about when it happens.

If one course fails, i think we should just fix the error, rather than continuing to sync other things.